### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Limits.Shapes.BinaryProducts

### PR DESCRIPTION
**Pull request by mattrobball**
---

- [x] depends on: #2496
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2496
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2496~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean [hunk 1]
```diff
@@ -0,0 +1,1437 @@
...
+
+instance {X : C} [HasBinaryProduct X X] : IsSplitMono (diag X) :=
+  IsSplitMono.mk' { retraction := prod.fst }
+
+end ProdLemmas
+
+section CoprodLemmas
+
+-- @[reassoc (attr := simp)]
+@[simp] -- Porting note: removing reassoc tag since result is not hygenic (two h's)
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean:824]: Has this been reported as `reassoc` bug?
Is it a viable workaround to rename the `h` in the statement to `h'`?

**Review by jcommelin — COMMENTED** [review]
1 comment, but LGTM

bors d+

File: Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean:824]: I started a [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/hygenic.20reassoc) topic but I couldn't find a prior mention.

**Comment by mattrobball** [comment]
bors merge

File: Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean:824]: Yeah, I guess renaming it would be easier.

**Pull request closed without merging**